text/3503-frontmatter.md

```diff
@@ -0,0 +1,662 @@
+~~~
+
+fn main() {
+}
+```
+```rust
+#!/usr/bin/env cargo
++++cargo
+[package]
+edition = "2018"
++++
+
+fn main() {
+}
+```
+```rust
+#!/usr/bin/env cargo
+===cargo
+[package]
+edition = "2018"
+===
+
+fn main() {
+}
+```
+```rust
+#!/usr/bin/env cargo
+>>>cargo
+[package]
+edition = "2018"
+<<<
+
+fn main() {
+}
+```
+```rust
+#!/usr/bin/env cargo
+<<<cargo
+[package]
+edition = "2018"
+>>>
+
+fn main() {
+}
+```
+
+Downsides
+- With `>>>` it isn't quite like HEREDOC to have less overhead
+- `>>>`, `<<<`, `|||`, `===` at the beginning of lines start to look like merge conflicts which might confuse external tools
+- Backticks have a problem with users knowing how to and remembering to escape these blocks when sharing them in markdown.  Knowing the syntax (only because I've implemented a parser for it), I'm at about 50/50 on whether I properly escape.
+
+Note:
+- `"` was not considered because that can feel too familiar and users might carry over their expectations for how strings work
+
+### Alternative 2: Extended Shebang
+
+````rust
+#!/usr/bin/env cargo
+# ```cargo
+# [dependencies]
+# foo = "1.2.3"
+# ```
+
+fn main() {}
+````
+This is a variation on other options that ties itself closer to the shebang syntax.
+The hope would be that we could get buy-in from other languages.
+- The first line post-shebang-stripping is a hash plus 3+ backticks, then capture all content until a matching pair of backticks on a dedicated line.  This would be captured into a `#![frontmatter(info = "cargo", content = "..."]`.  `frontmatter` attribute is reserved for crate roots.  The 3+ with matching pair is a "just in case" a TOML multi-line string has that syntax in it).  Each content line must be indented to at least the same level as the first backtick.
+  - Backticks are needed to know to avoid parsing `#[dependencies]` as an attribute
+  - This also allows an infostring so this isn't just a cargo feature
+- Future evolution: Allow `cargo` being the default `info` string
+- Future evolution: Allow any `info` string with cargo checking for `content.starts_with(["cargo", "cargo,"])`
+- Future evolution: Allow `frontmatter` attribute on any module
+
+If we dropped future possibilities for additional content, we could remove the opening/closing syntax and tweak it to avoid attribute ambiguity,
+greatly reducing the minimum syntax needed in some cases.
+````rust
+#!/usr/bin/env cargo
+## package.edition = "2018"
+
+fn main() {}
+````
+
+Benefits
+- Visually connected to the shebang
+- Has parallels to ideas outside of Rust, building on external knowledge that might exist
+- Easy for cargo to parse and modify
+- Can easily be leveraged by buck2, meson, etc in the future
+- Maybe we can get others on board with this syntax
+
+Downsides
+- More syntactically heavy than the frontmatter solution
+  - Visually
+  - More work to type it out / copy-paste
+  - More to get wrong
+- Requires users to deal with leading characters when editing/copying/pasting the manifest
+
```

Comment:
```suggestion
- `#` prefix plus a TOML `[heading]` looks too much like a Rust `#[attribute]`.

```

I don't find the other downsides compelling, and I would *otherwise* argue in favor of this solution, but I do find this counterargument sufficiently compelling.

Reply:
I added your line in and better highlighted this by pointing out an example with a table.

For myself, I think it is heavier syntactically and I think that should still be called out.  I was redundant on saying that though, so I removed the last line.  I also clarified what I meant by copy/paste.